src/user-languages.ts

```diff
@@ -1,132 +1,114 @@
-  try {
-    const vscode = getVscode();
-
-    const decoder = new TextDecoder("utf-8");
-    const extensionLangs: LanguageRegistration[] = [];
-    const seenScope = new Set<string>();
-
-    // Normalize langIds to lowercase for comparison
-    const normalizedLangIds = langIds?.map((id) => id.toLowerCase());
-
-    for (const extension of vscode.extensions.all) {
-      const contributes = (extension.packageJSON?.contributes ?? {}) as {
-        languages?: { id: string; aliases?: string[] }[];
-        grammars?: {
-          language?: string;
-          scopeName?: string;
-          path?: string;
-          embeddedLanguages?: Record<string, string>;
-        }[];
-      };
-
-      if (!contributes.grammars || contributes.grammars.length === 0) {
-        continue;
-      }
+import type { LanguageRegistration } from "shiki";
+import { LanguageRegistrationCollectionBuilder, type LanguageRegistrationExtended } from "./shiki-bridge-language.js";
+import { ExtensionFileReader, getVscode } from "./vscode-utils.js";
+import { LanguageRegistry } from "./language-registry.js";
+
+interface UserLangsResult {
+  /**
+   * The language registrations to pass to Shiki's highlighter.
+   */
+  langs: LanguageRegistration[],
+  /**
+   * Get the language registration for the given language id.
+   * Will resolve language id if it is an alias.
+   *
+   * Returns `undefined` if there is no language registration for the given language id.
+   *
+   * @example
+   * ```ts
+   * const result = getUserLangs(['tsx']);
+   *
+   * const resolvedLanguageId = result.get('tsx');
+   * //    ^? LanguageRegistration { name: 'typescriptreact', ... }
+   * ```
+   */
+  get(languageId: string): LanguageRegistration | undefined,
+  /**
+   * A helper function to resolve a possible alias to its language id.
+   * The language registrations always use the resolved alias as its `name` property.
+   * All its aliases can be found under the `aliases` property.
+   *
+   * @example
+   * ```ts
+   * const result = getUserLangs(['tsx']);
+   *
+   * const resolvedLanguageId = result.resolveAlias('tsx');
+   * //    ^? 'typescriptreact'
+   * ```
+   */
+  resolveAlias(languageId: string): string;
+  /**
+   * A helper function to resolve an `.ext` extension to its language id.
+   * @example
+   * ```ts
+   * const result = getUserLangs(['handlebars']);
+   *
+   * const resolvedLanguageId = result.resolveExtension('.hbs');
+   * //    ^? 'handlebars'
+   */
+  resolveExtension(extension: string): string;
+}
 
-      // Build language ID to aliases map
-      const languageIdToAliases = new Map<string, string[]>();
-      for (const lang of contributes.languages ?? []) {
-        if (lang?.id) languageIdToAliases.set(lang.id, lang.aliases ?? []);
-      }
+let cache: LanguageRegistry | null = null;
+function getLanguageRegistry(vscode: typeof import('vscode')): LanguageRegistry {
+  if (!cache) {
+    cache = LanguageRegistry.build(vscode.extensions.all);
```

Comment:
Same here, I think we should load the languages the user requested and their dependencies. I'm not sure it's feasible without loading every single one of them to the registry but it's worth trying

Reply:
I see, make sense